### PR TITLE
feat: api/cli: beneficiary withdraw api and cli

### PR DESCRIPTION
**Pull request by geoff-vball**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/9257 and https://github.com/filecoin-project/lotus/issues/9258

## Proposed Changes
Adds an api call and cli command to withdraw funds from a miner as the beneficiary.

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Review by arajasek — CHANGES_REQUESTED** [review]
I don't think we want the new API method in the FullNodeAPI here, just the Storage API. See `ActorWithdrawBalance` for an equivalent method to use as a blueprint.

**Comment by geoff-vball** [comment]
@arajasek Does the storage API get run on the `lotus` node or through `lotus-miner`? I think I feel pretty strongly that this is an action that shouldn't require running `lotus-miner` to achieve.

**Comment by arajasek** [comment]
@geoff-vball Yes, api_storage is lotus-miner only. I would match what we currently have for `(Owner)WithdrawBalance` for now (which is to only have it on the miner API), and create a ticket to support both `(Owner)WithdrawBalance` and `BeneficiaryWithdrawBalance` over the node API. I'd be...skeptical about whether these are actions we want to be able to take remotely over an API, but we can descope that discussion.

Additionally, I would move the CLI commands to the following two locations, you can match existing behaviour there: https://github.com/filecoin-project/lotus/blob/e65fae28de2a44947dd24af8c7dafcade29af1a4/cmd/lotus-shed/actor.go#L39, https://github.com/filecoin-project/lotus/blob/15ea3953f09311d1587d1c3291013771ca7ba3e7/cmd/lotus-miner/actor.go#L233. The first can be called by anyone who has built lotus-shed, the second requires the miner running in the background.

TL;DR: let's match what we have for the existing withdrawBalance method and take it from there.

File: cmd/lotus-shed/actor.go [hunk 1]
```diff
@@ -159,6 +160,78 @@ var actorWithdrawCmd = &cli.Command{
...
+		minerAPI, closer, err := lcli.GetStorageMinerAPI(cctx)
+		if err != nil {
+			return err
+		}
+		defer closer()
+
+		msgCid, err := minerAPI.BeneficiaryWithdrawBalance(ctx, maddr, amount)
+		if err != nil {
+			return err
+		}
```
Comment (geoff-vball) [inline cmd/lotus-shed/actor.go:207]: @arajasek I imagine that I might have to used `MpoolPushMessage` here instead of relying on the miner-api that might not be present?

File: cmd/lotus-miner/actor.go [hunk 2]
```diff
@@ -40,6 +40,7 @@ var actorCmd = &cli.Command{
 	Subcommands: []*cli.Command{
 		actorSetAddrsCmd,
 		actorWithdrawCmd,
```
Comment (arajasek) [inline cmd/lotus-miner/actor.go:42]: Can we rename this to ownerWithdrawCmd, rename the command itself to `owner-withdraw`, but leave `withdraw` as an alias?

File: cmd/lotus-miner/actor.go [hunk 3]
```diff
@@ -305,6 +306,78 @@ var actorWithdrawCmd = &cli.Command{
 	},
 }
 
+var beneficiaryWithdrawCmd = &cli.Command{
+	Name:      "beneficiary-withdraw",
+	Usage:     "withdraw available balance to beneficiary address. Will not work from non-beneficiary owner addresses. Use lotus-miner for non-beneficiary owner withdrawals",
```
Comment (arajasek) [inline cmd/lotus-miner/actor.go:311]: This helptext  is confusing me -- what are you trying to say here? I think what you need to tell the user is that this command will try to _send_ the message from the beneficiary address. If one isn't set, or if they just want to use the owner address, they should instead use `lotus-miner withdraw`?

File: cmd/lotus-shed/actor.go (hunk 1 above)
Comment (arajasek) [reply cmd/lotus-shed/actor.go]: Yes, see the command directly above this one. Honestly, you could fold this command into that one -- just add a param for the user to indicate whether to send from owner or beneficiary. They're otherwise identical.

File: node/impl/storminer.go [hunk 4]
```diff
@@ -1387,3 +1387,43 @@ func (sm *StorageMinerAPI) ActorWithdrawBalance(ctx context.Context, amount abi.
 
 	return smsg.Cid(), nil
 }
+
+func (sm *StorageMinerAPI) BeneficiaryWithdrawBalance(ctx context.Context, minerAddress address.Address, amount abi.TokenAmount) (cid.Cid, error) {
```
Comment (arajasek) [inline node/impl/storminer.go]: That's a lot of duplicated code -- let's extract all of this into a helper that takes a `useOwner` boolean param indicating who the sender should be?

File: cmd/lotus-miner/actor.go [hunk 5]
```diff
@@ -232,14 +233,18 @@ var actorSetPeeridCmd = &cli.Command{
...
 	ArgsUsage: "[amount (FIL)]",
 	Flags: []cli.Flag{
 		&cli.IntFlag{
 			Name:  "confidence",
 			Usage: "number of block confirmations to wait for",
 			Value: int(build.MessageConfidence),
 		},
+		&cli.BoolFlag{
+			Name:  "beneficiary",
+			Usage: "send withdraw command using beneficiary key",
```
Comment (arajasek) [inline cmd/lotus-miner/actor.go:246]:
```suggestion
			Usage: "send withdraw message from the beneficiary address",
```

File: cmd/lotus-miner/actor.go (hunk 2 above)
Comment (geoff-vball) [reply cmd/lotus-miner/actor.go]: Refactored everything. Now the cli commands just need a flag to know to send from beneficiary.

File: cmd/lotus-shed/actor.go [hunk 6]
```diff
@@ -50,6 +50,10 @@ var actorWithdrawCmd = &cli.Command{
 			Usage: "number of block confirmations to wait for",
 			Value: int(build.MessageConfidence),
 		},
+		&cli.BoolFlag{
+			Name:  "beneficiary",
+			Usage: "send withdraw command using beneficiary key",
```
Comment (arajasek) [inline cmd/lotus-shed/actor.go:55]:
```suggestion
			Usage: "send withdraw message from the beneficiary address",
```
